### PR TITLE
Add subcription approval capability to publisher portal

### PR DESCRIPTION
**Pull request by shilmyhasan**
This featured is intended to provide the subscription approval capability to the publisher via publisher portal. The feature works similar to the feature we already have in admin portal for subcription approval

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
+package org.wso2.carbon.apimgt.rest.api.publisher.v1.impl;
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:1]: Add license header

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+import org.wso2.carbon.apimgt.rest.api.publisher.v1.dto.*;
+import org.apache.cxf.jaxrs.ext.MessageContext;
+import org.wso2.carbon.apimgt.rest.api.publisher.v1.dto.WorkflowListDTO;
+import org.wso2.carbon.apimgt.rest.api.publisher.v1.utils.WorkflowMappingUtil;
+import org.wso2.carbon.apimgt.rest.api.util.utils.RestApiUtil;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import javax.ws.rs.core.Response;
+import java.util.*;
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:34]: Don't use wildcard imports

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+import static org.wso2.carbon.utils.multitenancy.MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:36]: Add new line after this

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+     * @param externalWorkflowRef is the unique identifier for workflow request
+     * @return
+     */
+    @Override
+    public Response workflowsExternalWorkflowRefGet(String externalWorkflowRef, MessageContext messageContext)
+            throws APIManagementException {
+        WorkflowInfoDTO workflowinfoDTO;
+        try {
+            Workflow workflow;
+            String status = "CREATED";
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:53]: Why creating a new variable here? Instead use a constant

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+     *
+     * @param externalWorkflowRef is the unique identifier for workflow request
+     * @return
+     */
+    @Override
+    public Response workflowsExternalWorkflowRefGet(String externalWorkflowRef, MessageContext messageContext)
+            throws APIManagementException {
+        WorkflowInfoDTO workflowinfoDTO;
+        try {
+            Workflow workflow;
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:52]: Move this out of the try block

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        offset = offset != null ? offset : RestApiConstants.PAGINATION_OFFSET_DEFAULT;
+        String tenantDomain = RestApiCommonUtil.getLoggedInUserTenantDomain();
+        String query = "";
+        Map<String, Object> result;
+        APIProvider apiProvider = RestApiCommonUtil.getLoggedInUserProvider();
+        boolean isMore = true;
+        int start = 0;
+        List<APIIdentifier> apiIdentifiers = null;
+        while(isMore) {
+            result = apiProvider.searchPaginatedAPIs(query, tenantDomain, 0, start+100, RestApiConstants.DEFAULT_SORT_CRITERION, RestApiConstants.DESCENDING_SORT_ORDER );
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:86]: Format code. ex `start+100` as `start + 100`

Comment (lasanthaS) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:86]: Format other places as well

**Review by lasanthaS — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            }
+        }
+        WorkflowListDTO workflowListDTO;
+        try {
+            Workflow[] workflows;
+            List<Workflow> workflowList;
+            String status = "CREATED";
+            APIAdmin apiAdmin = new APIAdminImpl();
+            if(workflowType != null) {
+                if (workflowType.equals("SUBSCRIPTION_CREATION")) {
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:104]: Move these to constants

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                    WorkflowUtils.sendNotificationAfterWFComplete(workflowDTO, workflowType);
+                }
+                APIUtil.logAuditMessage(APIConstants.AuditLogConstants.WORKFLOW_STATUS, new Gson().toJson(workflowDTO),
+                        APIConstants.AuditLogConstants.UPDATED, RestApiCommonUtil.getLoggedInUsername());
+                return Response.ok().entity(body).build();
+            } else {
+                RestApiUtil.handleBadRequest("Invalid Workflow Type", log);
+            }
+        } catch (APIManagementException e) {
+            String msg = "Error while resuming workflow " + workflowReferenceId;
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:195]: No need of assigning a new `msg` variable.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (lasanthaS) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java:2]: Year should be 2024 and LLC instead of Inc.

**Review by lasanthaS — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        while(isMore) {
+            result = apiProvider.searchPaginatedAPIs(query, tenantDomain, 0, start+100, RestApiConstants.DEFAULT_SORT_CRITERION, RestApiConstants.DESCENDING_SORT_ORDER );
+            apiIdentifiers  =  new ArrayList<>();
+            Set<API> apis = (Set<API>) result.get("apis");
+            for(API api : apis){
+                apiIdentifiers.add(api.getId());
+            }
+            isMore = Boolean.parseBoolean(result.get("isMore").toString());
+            if(isMore) {
+                start = start + 100;
```
[the remark this answers is not in the document]
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Fixed in 19c878193dc11ff73894e502d43a572b0844319c

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        List<WorkflowInfoDTO> workflowInfoDTOs = workflowListDTO.getList();
+
+        if (workflowInfoDTOs == null) {
+            workflowInfoDTOs = new ArrayList<>();
+            workflowListDTO.setList(workflowInfoDTOs);
+        }
+
+        //identifying the proper start and end indexes
+        int start = offset < workflows.length && offset >= 0 ? offset : Integer.MAX_VALUE;
+        int end = offset + limit - 1 <= workflows.length - 1 ? offset + limit - 1 : workflows.length - 1;
```
[the remark this answers is not in the document]
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java]: Fixed in 19c878193dc11ff73894e502d43a572b0844319c

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java (hunk 9 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java]: Fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 8 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 7 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: These string are used only one time in the class.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 5 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 1 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:19]: Refacored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 4 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 2 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+    @Override
+    public Response workflowsExternalWorkflowRefGet(String externalWorkflowRef, MessageContext messageContext) {
+        try {
+            String tenantDomain = RestApiCommonUtil.getLoggedInUserTenantDomain();
+            APIAdmin apiAdmin = new APIAdminImpl();
+            Workflow workflow = apiAdmin.getworkflowReferenceByExternalWorkflowReferenceID(externalWorkflowRef,
+                    createdStatus, tenantDomain);
+            return Response.ok().entity(WorkflowMappingUtil.fromWorkflowsToInfoDTO(workflow)).build();
+        } catch (APIManagementException e) {
+            RestApiUtil.handleInternalServerError("Error while retrieving workflow request by the " +
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:75]: avoid having catch and throw up with error code.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+                                                      MessageContext messageContext) {
+        ApiMgtDAO apiMgtDAO = ApiMgtDAO.getInstance();
+        boolean isTenantFlowStarted = false;
+        String username = RestApiCommonUtil.getLoggedInUsername();
+        String tenantDomainOfUser = MultitenantUtils.getTenantDomain(username);
+        try {
+            if (workflowReferenceId == null) {
+                RestApiUtil.handleBadRequest("workflowReferenceId is empty", log);
+            }
+            org.wso2.carbon.apimgt.impl.dto.WorkflowDTO workflowDTO = apiMgtDAO.retrieveWorkflow(workflowReferenceId);
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:180]: move business logic into APIPublisherimpl

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,232 @@
+/*
+ *
+ *  Copyright (c) 2024, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java:3]:
```suggestion
 *  Copyright (c) 2024, WSO2 LLc. (http://www.wso2.org) All Rights Reserved.
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
+/*
+ *  Copyright (c) 2024, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/WorkflowMappingUtil.java:2]:
```suggestion
 *  Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 16]
```diff
@@ -8681,6 +8681,153 @@ paths:
...
+        - $ref: '#/components/parameters/offset'
+        - $ref: '#/components/parameters/Accept'
+        - name: workflowType
+          in: query
+          description: |
+            We need to show the values of each workflow process separately .for that we use workflow type.
+            Workflow type can be AM_APPLICATION_CREATION, AM_SUBSCRIPTION_CREATION,   AM_USER_SIGNUP, AM_APPLICATION_REGISTRATION_PRODUCTION, AM_APPLICATION_REGISTRATION_SANDBOX.
+          schema:
+            type: string
+            enum:
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:8706]: We don't need all these enums for publisher

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 17]
```diff
@@ -8681,6 +8681,153 @@ paths:
...
+        This operation can be used to retrieve list of workflow pending processes.
+      parameters:
+        - $ref: '#/components/parameters/limit'
+        - $ref: '#/components/parameters/offset'
+        - $ref: '#/components/parameters/Accept'
+        - name: workflowType
+          in: query
+          description: |
+            We need to show the values of each workflow process separately .for that we use workflow type.
+            Workflow type can be AM_APPLICATION_CREATION, AM_SUBSCRIPTION_CREATION,   AM_USER_SIGNUP, AM_APPLICATION_REGISTRATION_PRODUCTION, AM_APPLICATION_REGISTRATION_SANDBOX.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:8703]: remove unnessary entries that are not related to publisher

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 18]
```diff
@@ -8955,6 +9102,108 @@ components:
...
+    WorkflowInfo:
+      title: Workflow info object with basic workflow details
+      type: object
+      properties:
+        workflowType:
+          type: string
+          description: |
+            Type of the Workflow Request. It shows which type of request is it.
+          example: APPLICATION_CREATION
+          enum:
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:9163]: remove unnecessary entries

**Review by chamilaadhi — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml (hunk 16 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:9089]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml (hunk 18 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:9538]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml (hunk 17 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 13 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Refactored

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java (hunk 12 above)
Comment (shilmyhasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/WorkflowsApiServiceImpl.java]: Refactored

**Review by tharindu1st — APPROVED** [review]

**Review by chamilaadhi — APPROVED** [review]

**Review by lasanthaS — APPROVED** [review]
